storage/filesystem/object.go

```diff
@@ -79,8 +137,23 @@ func (s *ObjectStorage) Get(h core.Hash) (core.Object, error) {
 func (s *ObjectStorage) Iter(t core.ObjectType) (core.ObjectIter, error) {
 	var objects []core.Object
 
+	_, hashes, err := s.dir.Objectfiles()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, hash := range hashes {
+		object, err := s.getFromObject(hash)
+		if err != nil {
+			return nil, err
+		}
+		if object.Type() == t {
+			objects = append(objects, object)
+		}
+	}
+
 	for hash := range s.index {
-		object, err := s.Get(hash)
+		object, err := s.getFromPackfile(hash)
```

Comment:
Creating an slice of decoded objects and then returning an iterator over it is very inefficient and wastes a lot of memory, it will be much better if the iterator would be able to decode each object on demand with each Next() invocation.

This is not your fault, it is something we will change in the future.